### PR TITLE
[To Do] fix bugs and optimize flow

### PR DESCRIPTION
**Pull request by feich-ms**
fix some bugs and optimize some flow
## Description
1. fix the flow bug of AddFirstTask in ShowToDoDialog to avoid tasks being added twice which is a critical bug for first time user
2. Add SwitchListType logic to optimize the flow. This is meaningful especially when users are adding a shopping or grocery item but they didn't realize that. For examaple, when users say "remind me to buy eggs". Before the change, the "eggs" will be added to the grocery list automatically and the users will not clearly understand the reason. After the change, the bot will prompt to users that "Alright, this looks like a task for your Grocery list. Would you like me to add this to your Grocery list?". If users say "yes", then it will be added to grocery list, otherwise, it will be added to the basic to do list.
3. Add list type(**To Do/Shopping/Grocery**) for all the responses which are not clear for users. For example, before the change, when users say "show my to do list", the adaptive card will show "You have # items on your list". After the change, it will be "You have # items on you **To Do** list". This is very meaningful when users add/mark/delete tasks.
4. Optimze some unit test logic to cover more code for service client.

## Related Issue
#499

## Testing Steps
See descriptions

## Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the appropriate unit tests

### CODE REVIEW

**Comment by feich-ms** [comment]
@lzc850612 Hi Ted, I have reverted the changes related to "ReplaceDialogAsync". This PR is fixing some critical bugs of some corner cases. Please help to review. Thank you very much.

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs [hunk 1]
```diff
@@ -22,13 +22,13 @@ public void Initialize()
...
             var service = new OneNoteService();
-            var pageId = new Dictionary<string, string>() { { "ToDo", "ToDo" }, { "Grocery", "Grocery" }, { "Shopping", "Shopping" } };
+            var pageId = new Dictionary<string, string>();
 
             await service.InitAsync("test", pageId, mockClient);
 
-            var taskList = await service.GetTasksAsync("ToDo");
-            await service.AddTaskAsync("ToDo", "Test 9");
-            var addedTaskList = await service.GetTasksAsync("ToDo");
+            var taskList = await service.GetTasksAsync("To Do");
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs]: better define this string other strings used as const somewhere and reference that

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs (hunk 1 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs]: Fixed. I replaced all the hard-coded "To Do" with ToDoStrings.ToDo which comes from resouce file used in ToDoSkill project

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/Fakes/MockHttpClientHandlerGen.cs [hunk 2]
```diff
@@ -110,11 +122,33 @@ private void SetHttpMockBehavior(ref Mock<HttpClientHandler> mockClient)
...
+              {
+                  Content = new StringContent(this.GetPageDetails()),
+              });
+
             mockClient
                 .Protected()
                 .Setup<Task<HttpResponseMessage>>(
                 "SendAsync",
-                ItExpr.Is<HttpRequestMessage>(r => r.RequestUri.ToString().StartsWith("https://graph.microsoft.com/beta/me/outlook/taskFolders/ToDo/tasks") && r.Method == HttpMethod.Get),
+                ItExpr.Is<HttpRequestMessage>(r => r.RequestUri.ToString().StartsWith("https://graph.microsoft.com/beta/me/outlook/taskFolders/To Do/tasks") && r.Method == HttpMethod.Get),
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/Fakes/MockHttpClientHandlerGen.cs:150]: https://graph.microsoft.com/beta/me/outlook/taskFolders/To Do/tasks

'To Do' here is separated. is this right?

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs [hunk 3]
```diff
@@ -52,13 +53,13 @@ public async Task MarkTaskTest()
...
             var service = new OneNoteService();
-            var pageId = new Dictionary<string, string>() { { "ToDo", "ToDo" }, { "Grocery", "Grocery" }, { "Shopping", "Shopping" } };
+            var pageId = new Dictionary<string, string>();
 
             await service.InitAsync("test", pageId, mockClient);
 
-            var taskList = await service.GetTasksAsync("ToDo");
-            await service.DeleteTasksAsync("ToDo", taskList.GetRange(0, 1));
-            var deletedTaskList = await service.GetTasksAsync("ToDo");
+            var taskList = await service.GetTasksAsync(ToDoStrings.ToDo);
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs]: I don't think you should add 'ToDo' to resource strings. Adding it to resource strings means that it changes per culture. But this has to be always 'Todo' in all languages right?

You should just create a const string somewhere and just reference that from all the places you want to use it.

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/Flow/AddToDoFlowTests.cs [hunk 4]
```diff
@@ -175,7 +182,7 @@ private Action<IActivity> ShowUpdatedShoppingList()
                 Assert.IsNotNull(toDoChoices);
                 var toDoChoiceCount = toDoChoices.Items.Count;
                 CollectionAssert.Contains(
-                    this.ParseReplies(ToDoSharedResponses.ShowToDoTasks.Replies, new StringDictionary() { { "taskCount", (MockData.MockShoppingItems.Count + 1).ToString() } }),
+                    this.ParseReplies(ToDoSharedResponses.ShowToDoTasks.Replies, new StringDictionary() { { "taskCount", (MockData.MockShoppingItems.Count + 1).ToString() }, { "listType", "Shopping" } }),
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/Flow/AddToDoFlowTests.cs]: "listType", "shopping" should also be defined in const strings and reference them from here.

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs (hunk 3 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs]: The principal should be that if you have to use a hard-coded string from multiple places (actually even if you're only going to use it once), you should define that string somewhere and reference that variable in the places you want to use it. This way if in the future you have to change the string, you only have to change one place.

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/Fakes/MockHttpClientHandlerGen.cs (hunk 2 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/Fakes/MockHttpClientHandlerGen.cs:150]: Yes. It is right.

Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/Fakes/MockHttpClientHandlerGen.cs:150]: ok

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs (hunk 3 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/API/OneNoteServiceTests.cs]: fixed. I replaced all the hard-coded strings which are used more once with const string parameters.

File: solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/Flow/AddToDoFlowTests.cs (hunk 4 above)
Comment (feich-ms) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/todoskilltest/Flow/AddToDoFlowTests.cs]: fixed. I replaced all the hard-coded strings which are used more once with const string parameters.

**Review by lzc850612 — APPROVED** [review]
